src/openforms/submissions/form_logic.py

```diff
@@ -182,21 +184,22 @@ def evaluate_form_logic(
         # debounced client-side data changes
         data_diff = {}
 
-        for key, new_value in step.data.items():
-            original_value = initial_data.get(key)
+        components = iter_components(configuration)
+        for key in map(lambda x: x["key"], components):
```

Comment:
```suggestion
        for (key := component["key"]) in iter_components(configuration):
```

I don't understand why we can't loop over `step.data.items()` though. As far as I can see, we get the merged object of (saved data + dirty data), where dirty data overwrites the saved data if keys collide.

Reply:
From my observations `step.data` often has missing keys, and here we have to iterate over them.

Reply:
Anyhow I'd like to discuss this function in more detail. It is doing too much at once and will be a primary source for bugs. The logic of mutation operations needs a revision.

Reply:
> From my observations step.data often has missing keys, and here we have to iterate over them.

This should be made clear by adding comments to the code.

> Anyhow I'd like to discuss this function in more detail. It is doing too much at once and will be a primary source for bugs. The logic of mutation operations needs a revision.

This function has been refactored only a week ago with more separation of concerns. "it will be a primary source of bugs"  -> you know we write tests to prevent those, right? Those are unfounded claims.

There are business requirements and technical requirements from the libraries we use. This function does a lot, but it is well documented why things are the way they are and it is well-tested.

Reply:
Good to know it is already being refactored :)

I agree writing tests is an excellent to prevent bugs, note that it is unrelated to my point about following simple principles of clean architecture.